### PR TITLE
[fixed] Storage for Aliyun Block Storage

### PR DESCRIPTION
**Pull request by thxCode**
1. change default created diskCatagory (cloud -> cloud_ssd) and size(5 -> 20)
2. change script logic
3. enforce judgment

### CODE REVIEW

**Comment by thxCode** [comment]
@Tracy1994 ,
[#9807](https://github.com/rancher/rancher/issues/9807)
[#9808](https://github.com/rancher/rancher/issues/9808)

**Comment by thxCode** [comment]
@Tracy1994 ,
here is the [testing catalog](https://github.com/thxCode/rancher-catalog/tree/aliyun-bs).

**Comment by yasker** [comment]
@Tracy1994 please test.

**Comment by yasker** [comment]
Also @thxCode
1. Please add a link to original general issue, otherwise, we would lose track of this PR: https://github.com/rancher/rancher/issues/9194
2. And let me know after you submit the PR and ready to be reviewed.

File: package/abs/rancher-abs [hunk 1]
```diff
@@ -187,15 +187,16 @@ detach_aliyun_disk() {
 init() {
     get_instance_id
 
-    log "\n[init] on ${TRANSCATION_SEQ}" /tmp/rancher_abs.log
+    rm -f /tmp/rancher_abs.log
```
Comment (yasker) [inline package/abs/rancher-abs:224]: Please separate your debug log changes into a separate commit.

File: package/abs/rancher-abs [hunk 2]
```diff
@@ -187,15 +187,16 @@ detach_aliyun_disk() {
...
+    log "\n[init] on ${TRANSCATION_SEQ}"
 
     local region_id=$(curl http://100.100.100.200/latest/meta-data/region-id 2>/dev/null)
```
Comment (yasker) [inline package/abs/rancher-abs]: You have multiple places to retrieve region_id and zone_id. Put them into reuseable functions.

File: package/abs/rancher-abs [hunk 3]
```diff
@@ -395,24 +393,37 @@ mountdest() {
...
+        # for DiskCatagory: cloud
         if [ "$(echo ${mount_result} | grep 'wrong fs type')" ]; then
-            log "> fail, try to fix. ${mount_result}" /tmp/rancher_abs.log
+            try_fix=1
+        # for DiskCatagory: cloud_ssd
+        elif [ "$(echo ${mount_result} | grep 'unknown filesystem type')" ]; then
+            try_fix=1
+        fi
+
+        if [ "${try_fix}"x = "1"x ]; then
```
Comment (yasker) [inline package/abs/rancher-abs]: Is there a possibility that you would format a disk contained valid data but you simply don't have the driver for it at the moment?

**Review by yasker — CHANGES_REQUESTED** [review]
The commit needs to be separated.

Also, there are issues with the original commit. I will review them in detail after the commit separation is done.

File: package/abs/rancher-abs (hunk 3 above)
Comment (gitlawr) [reply package/abs/rancher-abs]: It would be possible.
I think there are two problems here,
1.  only some certain types of error are captured here. For example, when 'device does not exist' error occurs the mount function ignores the error and returns a success.
2. We should do the fs formatting on volume creation instead of mounting.

File: package/abs/rancher-abs [hunk 4]
```diff
@@ -35,41 +37,48 @@ gen_disk_name() {
...
+}
 
 # - # https://help.aliyun.com/document_detail/25514.html?spm=5176.doc25516.6.868.8l8Kr9
 get_attached_device_path() {
-    local aliyun_options="--DiskIds '[\"${DISK_ID}\"]'"
-    local aliyun_call_result=$(aliyuncli ecs DescribeDisks "${aliyun_options}" --output json --filter Disks.Disk[0].Description 2>&1)
+    get_region_id
+
+    local aliyun_options="--DiskType data --RegionId ${REGION_ID}"
+    local aliyun_call_result=$(aliyuncli ecs DescribeDisks ${aliyun_options} --output json --filter Disks 2>&1)
```
Comment (gitlawr) [inline package/abs/rancher-abs]: Need to split the local declaration because it prevents error handling with $?

File: package/abs/rancher-abs [hunk 5]
```diff
@@ -124,48 +133,73 @@ listening_aliyun_disk() {
...
 # - # https://help.aliyun.com/document_detail/25515.html?spm=5176.doc25513.6.868.xKltry
 attach_aliyun_disk() {
-    log "\t> [disk attach]" /tmp/rancher_abs.log
+    log "\t> [disk attach]"
 
     local delete_with_instance=${1:-"false"}
 
     local aliyun_options="--InstanceId ${INSTANCE_ID} --DiskId ${DISK_ID} --DeleteWithInstance ${delete_with_instance}"
     local aliyun_call_result=$(aliyuncli ecs AttachDisk ${aliyun_options} --output json 2>&1)
     if [ $? -ne 0 ]; then
```
Comment (gitlawr) [inline package/abs/rancher-abs]: Take consideration of the case when the volume is attached to another host here, need to do a `detach` to make the volume available. Regarding issue https://github.com/rancher/rancher/issues/9820

File: package/abs/rancher-abs [hunk 6]
```diff
@@ -35,41 +37,48 @@ gen_disk_name() {
...
-        aliyun_call_result=$(echo "${aliyun_call_result}" | sed "s/\"//g")
-    fi
-
-    if [ "$(echo ${aliyun_call_result} | grep 'rancher_abs')" ]; then
-        aliyun_call_result=${aliyun_call_result#'/rancher_abs/'}
-    else
-        aliyun_call_result=""
-    fi
-
+    aliyun_call_result=$(echo "${aliyun_call_result}" | jq " .Disk[] | select(.DiskId==\"${DISK_ID}\") | .Description" | sed 's/"//g')
```
Comment (gitlawr) [inline package/abs/rancher-abs]: Merely filter by disk_id here may result in a device path in another host.

File: package/abs/rancher-abs [hunk 7]
```diff
@@ -35,41 +37,48 @@ gen_disk_name() {
...
+    # fi
+    aliyun_call_result=${aliyun_call_result#rancher_abs}
     echo "${aliyun_call_result}"
 }
 
 
 # Find avaiable device path in os, such as '/dev/vdb', '/dev/vdc',..., '/dev/vdz'
 find_available_device_path() {
-    local last_attach_device=$(lsblk -o KNAME | sed -e 's/^/\/dev\//' | awk 'END {print}')
+    local last_attach_device=$(lsblk -o KNAME,MOUNTPOINT | sed -e 's/^/\/dev\//' | awk '{if($2 ~ /^\w*$/){print $1}}' | awk 'END{print}')
```
Comment (gitlawr) [inline package/abs/rancher-abs]: I wonder if it makes sense to get a `last` device path as the available one.

**Comment by gitlawr** [comment]
I'll do follow-up changes to handle mentioned issues.

**Comment by gitlawr** [comment]
I've updated the PR.
@yasker PTAL
test catalog:
```
repo: https://github.com/gitlawr/rancher-catalog.git
branch: abs
```

**Comment by gitlawr** [comment]
And fix https://github.com/rancher/rancher/issues/9820

**Comment by yasker** [comment]
@gitlawr 

It's good to make incremental changes, but you don't need comments on your own commits. It's confusing. Also, you don't need to retain @thxCode 's original commits when it doesn't complete all the fixes.

Just create new incremental commits and address all your own comments.

Also, make titles more specific than `fix bugs` / `update attach`. The reader of the commit message won't get anything from reading the title. And it's better to include the subsystem name `ABS` in the title since the repo is for all different kinds of storage drivers.

**Comment by cjellick** [comment]
Very stale. Closing. Thanks! Please open a new PR if there is still a relevant change.
